cli/urfave/mod.go

```diff
@@ -0,0 +1,165 @@
+// Package urfave provides a cli builder implementation based on the urfave/cli
+// library.
+package urfave
+
+import (
+	"fmt"
+
+	ucli "github.com/urfave/cli/v2"
+	"go.dedis.ch/dela/cli"
+)
+
+// Builder implements a cli builder based on urfave/cli
+//
+// - implements cli.Builder
+type Builder struct {
+	commands map[string]*cmdBuilder
+	name     string
+	action   cli.Action
+	flags    []cli.Flag
+}
+
+// NewBuilder returns a new initialized builder. Action allows one to define a
+// primary action, but can be nil if we only needs to define commands. Flags
+// provides the global flags available from all the commands/subcommands.
+func NewBuilder(name string, action cli.Action, flags ...cli.Flag) cli.Builder {
+	return &Builder{
+		name:     name,
+		commands: make(map[string]*cmdBuilder),
+		action:   action,
+		flags:    flags,
+	}
+}
+
+// Build implements cli.builder.
+func (b Builder) Build() cli.Application {
+	app := &ucli.App{
+		Name:     b.name,
+		Commands: buildCommand(b.commands),
+		Action:   makeAction(b.action),
+		Flags:    buildFlags(b.flags),
+	}
+
+	app.Setup()
+
+	return app
+}
+
+// SetCommand implements cli.Builder.
+func (b *Builder) SetCommand(name string) cli.CommandBuilder {
```

Comment:
I'm a tiny bit conflicted.
In Go, I believe the best practice is to take abstractions as parameters and return concretions, i.e. take in interfaces and return structs. On the other hand, it makes a lot of sense to return that abstraction here, in terms of code eloquence...

Reply:
That would be useful only if we needed to use that function from within the package and we needed to handle a `cmdBuilder`, which is not the case. An external user would see the abstraction anyway.

Reply:
agreed.
I was merely referring to the practice of returning concretions and letting the client code decide on the interface that suits its needs. In this case, I agree that it doesn't make much sense.